src/platforms/stm32/src/lib/gpio_driver.c

```diff
@@ -222,6 +223,9 @@ void isr_error_handler(const char *isr_name);
 
 static term create_pair(Context *ctx, term term1, term term2)
 {
+    if (UNLIKELY(memory_ensure_free(ctx, TUPLE_SIZE(2)) != MEMORY_GC_OK)) {
+        RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+    }
```

Comment:
This change might trigger GC, making invalid terms in the caller. In this specific case this changes will break the following code:
```c
951         term ref = term_get_tuple_element(msg, 1);
952         ret_msg = create_pair(ctx, ref, ret);
 ```

When doing this kind of changes I suggest double checking, but also in general renaming a function to "foo_maybe_gc" might help understanding the side effects and double checking while doing find & replace (that should be done manually).

Reply:
Also, I think this change might be out of the scope of this PR.

Reply:
I had added that change at @fadushin's request. I can remove the `memory_ensure_free`, but I would like to fix this in a separate PR. Just to make sure I understand correctly, the check for `memory_ensure_free` should happen in the calling functions, before the use of `create_pair`, to avoid the kind of breakage you pointed out here?

Reply:
I reverted this change.

Reply:
> Just to make sure I understand correctly, the check for `memory_ensure_free` should happen in the calling functions, before the use of `create_pair`, to avoid the kind of breakage you pointed out here?

I looked a little closer and found the answer to this question. I will submit that PR soon.